### PR TITLE
feat: pod restart monitor for detecting broken daemonsets

### PR DESCRIPTION
**Pull request by iamnoah**
Hi, I wrote a really simple daemon to solve an issue we are having with the aws-node daemon restarting frequently and making nodes unstable. I'd love feedback on how to make it better and contribute.

### CODE REVIEW

**Comment by iamnoah** [comment]
/remove-lifecycle stale

Is this project dead?

File: pkg/podrestartmonitor/README.md
Comment (mmiranda96) [inline pkg/podrestartmonitor/README.md:14]: Already mentioned, but I think the pod selector should not have a default value.

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (mmiranda96) [inline pkg/podrestartmonitor/pod_restart_monitor.go:146]: What happens when we want to monitor more than one pod (either by having a broad `PodSelector` or configuring multiple monitors)? It looks like the condition will collide. Is there a workaround to prevent this?

**Review by mmiranda96 — CHANGES_REQUESTED** [review]
Thank you for your contribution @iamnoah! I've reviewed your change. Some comments are just questions and/or suggestions, other are hard requests. Let me know if you have any questions.

File: builder/Dockerfile
Comment (iamnoah) [reply builder/Dockerfile:15]: The go.mod file specifies `go 1.15`. Trying to build docker images without this change fails with multiple "module requires Go 1.15" errors.

File: Makefile
Comment (iamnoah) [reply Makefile:261]: I don't remember exactly, but I think I added this so I could override the value of ALL_BINARIES and build a docker image in a reasonable time. I can't get it to build at all now, build-in-docker just hangs when I try make.

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (iamnoah) [reply pkg/podrestartmonitor/pod_restart_monitor.go:158]: I don't have a good answer because I don't have that use case. I suppose the condition type could be configurable.

Comment (mmiranda96) [reply pkg/podrestartmonitor/pod_restart_monitor.go:162]: Looking at [official docs](https://kubernetes.io/docs/concepts/architecture/nodes/#condition), it appears that we can use reason for this. Let's keep `condition.Type` equal across all pods, but `condition.Reason` can contain the pod name to distinguish.

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (2rs2ts) [inline pkg/podrestartmonitor/pod_restart_monitor.go:92]: Although it would make the process of filtering pods more complex, it sure would be nice to be able to filter by annotations or by priority classes, in addition to labels. Feel free to ignore my comment here if you think it'd be a pain to implement, though.

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (2rs2ts) [inline pkg/podrestartmonitor/pod_restart_monitor.go:136]: I think this comment, and the name `l`, are relics of this being copied from the log monitor package. It might be good to update these, or at least, the comment 😅

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (2rs2ts) [reply pkg/podrestartmonitor/pod_restart_monitor.go:147]: I believe it will be an empty list. Try doing `kubectl get pod -l fakelabel=fakevalue -o json` in your cluster, you will get this:

```json
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```

So the loop will just do nothing, I think. It might trip up some people who get their namespace settings wrong or typo the selector when configuring their checks, but it would be consistent with how `kubectl` works.

**Review by 2rs2ts — COMMENTED** [review]
Not a maintainer, just a user with a keen interest in this leaving some thoughts. I've written a custom plugin monitor that does something similar, but am always keen to delete custom code in favor of official, bundled code, so I'd love to see this PR merged.

File: pkg/podrestartmonitor/pod_restart_monitor.go
Comment (mmiranda96) [reply pkg/podrestartmonitor/pod_restart_monitor.go:147]: That sounds good. I was worried that this might return an error, but an empty list is a handleable output.

For misconfigured selectors, we could have a warning log. But I don't see much value, I think it's good as is.

**Comment by sb3lbx** [comment]
Until this is merged, are there suitable alternatives?  Either available custom plugins or solutions outside of node-problem-detector?

I ask because our `aws-node` pods seem to randomly timeout on health checks.  In an effort to steady them, we're locking the resource requests and limits to tell Kubernetes to ensure the resources are always available.  However, it'd be real nice to have a solution that would taint our nodes where `aws-node` restarts.  Every time it restarts it breaks all workloads on that node.  The issue persists so we have to manually clean it up which is contrary to the idea of a "self-healing platform".

**Comment by 2rs2ts** [comment]
@sb3lbx although I am not at liberty to share the code, I can at least say that at my company, we worked around this by writing a custom plugin monitor that runs a bash script that checks the readiness state, and the number of container restarts in a rolling window, of any pod marked with a certain annotation. (In our case, we use the same annotation as the one we chose for [nodetaint](https://github.com/wish/nodetaint).) We made the thresholds for failure configurable because we found that we ended up needing to tweak how many restarts were a problem before the node should actually be considered unhealthy.

Maybe someday, if this PR dies of inactivity, I will look into whether we can make an open-source version of our script and contribute it back to this repo. I'd have to clear it with my company's open sourcing department, though, and it'd be better if it wasn't a custom plugin but rather its own kind of check, so I really do hope the owner of this PR comes back soon.

**Comment by sb3lbx** [comment]
Brilliant.  Thank you, @2rs2ts

**Pull request closed without merging**
